### DISCUSSION TITLE
ReferenceSprinklers - no automation

### DISCUSSION THREAD

**Discussion by aLAN-LDZ** [HomeSpan/HomeSpan#190] [Q&A]
Hi. Automation cannot be set in the home app. There is no accessory there. Is that an apple oversight?

**Comment by HomeSpan** [comment]
From what I've read, this is by design.  Automation routines for a sprinkler system need to be embedded in the device itself, with some form of interface for setting up the schedules.  I think the reason is that any automation routines that are operated by the Home App are subject to interruption, loss of WiFi, etc., that could leave the water running if the Home App can't send commands to the device (which is unfortunately not that uncommon for HomeKit).  Of course one could imagine simple time-out logic embedded in the device could mitigate this risk (e.g. the device shuts off all water if it does not receive a command from HomeKit in 30 mins, etc).  

The other reason for the lack of Home App sprinkler automation could be that most manufactures already have their own apps to set schedules that are specific to their device.  Apple trying to come up with a flexible, generic interface in the Home App that is better than a manufacturer-specific interface would likely be more trouble than it's worth.  So Apple may have made the decision that the primary use case for sprinklers in the Home App is to allow the user to manually trigger watering in a specific zone.  I think this is the purpose of the ProgramMode Characteristic.  It gets set by the device to tell the Home App user that the device is currently operating in "auto" mode, which in principal could override a manual setting.

For DIY systems, building a scheduling system into a HomeSpan sketch should be straightforward, especially if you add logic to set the time using a time server (something I probably should add as a native function of HomeSpan - maybe this is already the ESP32 IDF?).  The "harder" part is implementing an interface to control the scheduling.  One method would be via web pages hosted by the device.  It is possible to run a Web Server within HomeSpan (see my ProgrammableHub example under Projects), which would allow you to create a custom scheduling interface.  Another possibility is adding a display to the ESP32 for local programming.

It may also be possible to use Apple Shortcuts in conjunction with the Home App to create a sort-of schedule, but in my experience this is less robust - scheduling sometimes works, sometimes does not.

**Reply by aLAN-LDZ** [reply]
I modified the title sketch and And Basically it works, but if i want to add this to the bridge sketch, the program won't compile

main.ino:

`  new SpanAccessory();  
    new DEV_Identify("Irrigation","HomeSpan","123-ABC","XXX.XXX","0.9",0);
    new Head();`

Error:

`garden.ino: In function 'void setup()':
garden:116:14: error: no matching function for call to 'Head::Head()'
     new Head();
              ^
In file included from garden.ino:44:0:
sketch/Dev_Valve.h:43:3: note: candidate: Head::Head(const char*)
   Head(const char *headName) : Service::Valve() {
   ^
sketch/Dev_Valve.h:43:3: note:   candidate expects 1 argument, 0 provided
sketch/Dev_Valve.h:34:8: note: candidate: Head::Head(const Head&)
 struct Head : Service::Valve {
        ^
sketch/Dev_Valve.h:34:8: note:   candidate expects 1 argument, 0 provided
sketch/Dev_Valve.h:34:8: note: candidate: Head::Head(Head&&)
sketch/Dev_Valve.h:34:8: note:   candidate expects 1 argument, 0 provided
no matching function for call to 'Head::Head()'
`

**Comment by HomeSpan** [comment]
Note the constructor for Head is `Head(const char *headName)`.  You did not include a name in `new Head()` and thus the error messages.
